### PR TITLE
Update capitalization for iterator traits

### PR DESCRIPTION
**Pull request by iamed2**
For 0.7 compat

### CODE REVIEW

**Comment by ararslan** [comment]
I think these are available in Compat.

**Comment by iamed2** [comment]
Compat's version doesn't allow dispatch on `IteratorSize` on 0.6

**Comment by iamed2** [comment]
Ugh apparently this doesn't work

**Comment by iamed2** [comment]
Do I have to `import` them _and_ make the `const` aliases?

File: src/IterTools.jl [hunk 1]
```diff
@@ -5,8 +5,20 @@ module IterTools
...
-import Base: iteratorsize, IteratorSize, SizeUnknown, IsInfinite, HasLength, HasShape
-import Base: iteratoreltype, IteratorEltype, HasEltype, EltypeUnknown
+import Base: SizeUnknown, IsInfinite, HasLength, HasShape
+import Base: HasEltype, EltypeUnknown
+
+@static if VERSION < v"0.7.0-DEV.3309"
+    const IteratorSize = Base.iteratorsize
+    const IteratorEltype = Base.iteratoreltype
+    const IteratorSizeT = Base.IteratorSize
+    const IteratorEltypeT = Base.IteratorSize
```
Comment (ararslan) [inline src/IterTools.jl]: Is this right?

File: src/IterTools.jl [hunk 2]
```diff
@@ -5,8 +5,20 @@ module IterTools
...
+else
+    const IteratorSize = Base.IteratorSize
+    const IteratorEltype = Base.IteratorEltype
+    const IteratorSizeT = Base.IteratorSize
+    const IteratorEltypeT = Base.IteratorEltype
```
Comment (ararslan) [inline src/IterTools.jl]: I think for these you could just do `import Base: IteratorSize, IteratorEltype`? I'm not familiar with the `T` suffix for these two.

Comment (iamed2) [reply src/IterTools.jl]: That's something I just added to differentiate the type from the function.

The main problem is that I need to add methods to the function in most cases but in a couple cases I'm actually referencing the type. There's no way on 0.6 to use one thing for both.

File: src/IterTools.jl (hunk 1 above)
Comment (iamed2) [reply src/IterTools.jl]: Did I mean to do this? Yes. Does it work? No.

File: src/IterTools.jl (hunk 2 above)
Comment (iamed2) [reply src/IterTools.jl]: I did this right at the end of the day and it's messy and there's an obvious simpler way that I'll do tomorrow.

File: src/IterTools.jl (hunk 1 above)
Comment (iamed2) [reply src/IterTools.jl]: Update: I did not mean to do this and I misunderstood what you meant by "this"

**Comment by ararslan** [comment]
What I'd recommend:
```julia
@static if VERSION < v"0.7.0-DEV.3309"
    const _iteratorsize = Base.IteratorSize
    const _iteratoreltype = Base.IteratorEltype
else
    const _iteratorsize = Base.iteratorsize
    const _iteratoreltype = Base.iteratoreltype
end
```
That way you should be able to safely overload `_iteratorsize` and `_iteratoreltype`, and the types themselves can be referenced as `Base.IteratorSize` and `Base.IteratorEltype` across versions.

**Comment by iamed2** [comment]
My latest approach works once I fixed the typo. What do you think @ararslan ?

As a side note, it's pretty annoying to work with two versions of Julia on one package right now.

**Comment by ararslan** [comment]
>  What do you think @ararslan?

If it works for you and it works for CI, it works for me. 🙂

File: src/IterTools.jl [hunk 3]
```diff
@@ -33,7 +41,7 @@ function has_length(it)
 end
 
 promote_iteratoreltype(::HasEltype, ::HasEltype) = HasEltype()
-promote_iteratoreltype(::IteratorEltype, ::IteratorEltype) = EltypeUnknown()
+promote_iteratoreltype(::IteratorSize, ::IteratorSize) = EltypeUnknown()
```
Comment (martinholters) [inline src/IterTools.jl]: Huh?

**Comment by martinholters** [comment]
Just came across this after seeing an annoying spew of deprecation warnings... So I have no idea about the inner workings of this package, but couldn't you use `Compat.IteratorSize` for the function and `Base.IteratorSize` for the type? Of course, your current status is perfectly working, but re-introducing `iteratorsize` seems so backwards...

File: src/IterTools.jl (hunk 3 above)
Comment (iamed2) [reply src/IterTools.jl]: Alright, I'm officially not allowed to write open source PRs in a rush at the end of the day

**Comment by iamed2** [comment]
I think the existing approach is pretty simple; it just replicates the old environment so no functions or types are changed. 

Will merge soon.

File: src/IterTools.jl (hunk 3 above)
Comment (martinholters) [reply src/IterTools.jl]: Here's the bad news: It wasn't even Friday 😄

**Pull request merged**